### ISSUE TITLE
Thanks and some requests

### ISSUE THREAD

**Issue by san-r** [erickutcher/httpdownloader#49]
Thanks for creating a bloat-free download manager that can easily get download links (along with referer and cookies) from Firefox and Chrome browsers. Implementing communication between browser and download manager by running a server was a brilliant idea.

For a super light-weight program, it has almost all essential features. But if you find time, please consider the following additional requests:

- In the **"Add URL(s)"** dialog, please create a button **"Fetch size"** which would fetch and show the file size before starting download. When adding multiple urls, this might get complicated - so maybe enable this button only in case of a single url.
- Currently, the **"Add URL(s)"** dialog contains buttons for "Download" and "Cancel". Please add one more button **"Queue"**, which would just add to the queue of download tasks instead of downloading immediately.
- It already has support for http/https proxy. How about adding support for SOCKS 5 proxy? Another minimalist downloader MiniGet has this feature although it lacks many other features supported by HTTP downloader.

Issues that have been raised before by others but immensely useful to me as well:
- append an extension to incomplete file
- automatic rename in case file with same name exists

I think that the program would still be super light-weight after adding the above features.

Thanks again for this software.

**Comment by erickutcher** [comment]
Getting an accurate file size is a bit complicated and doesn't seem like something that should belong in the Add URL(s) window. There's a lot of steps involved (code-wise) and it could hang up on bad connections. However, you could get something similar by checking the "Simulate download" checkbox in the advanced settings of the Add URL(s) window. Once the downloads start you can just stop them. The process would be exactly the same, but this would allow you do see the file size for every URL that's added.

As far as queuing goes, that's determined by the "Active download limit" in the Options. If anything, I would have a button that adds the download in a paused state.

I've looked into SOCKS5 in the past, but not for this program. That would take some time to implement. I don't know the protocol, but I like the idea.

While creating the last version, I debated whether to use a file extension for incomplete files, or a temporary directory. I settled on a temporary directory. It's not enabled by default. You'll need to go into the Advanced options in the Options window and scroll to the bottom. Put a check in "Use temporary download directory".

Automatic rename can also be enabled in the Advanced options. Look for the setting called "When a file already exits" and set the drop down menu to "Rename File".

I'll keep these ideas in mind.

**Comment by san-r** [comment]
I regret missing automatic rename feature. And thanks for considering these points.

**Comment by erickutcher** [comment]
I've begun implementing SOCKS5 support and will try to get SOCKS4 working as well.

I tested it out with a random proxy I found and I was able to download a 100MB test file faster through the proxy than without it. Apparently my ISP's routing sucks.

**Comment by erickutcher** [comment]
The latest version is up and now supports SOCKS 4/4a/5.

**Comment by san-r** [comment]
I run a SOCKS5 proxy server (among others) in my workplace using 3proxy from https://3proxy.ru/ to share internet with other computers on LAN. That is, other computers on LAN can access internet through that SOCKS5 proxy server. I tested HTTP Downloader on a client machine and it downloaded files perfectly through that SOCKS5 server. Great job.

One question though. Other download managers like Flareget from https://flareget.com/ allow to enable only one proxy at a time. HTTP Downloader on the other hand allows all the three proxy (HTTP, HTTPS, SOCKS4/SOCKS5) options to be checked simultaneously. So I guess it will try sequentially?

**Comment by san-r** [comment]
I run a SOCKS5 proxy server (among others) in my workplace using 3proxy from https://3proxy.ru/ to share internet with other computers on LAN. That is, other computers on LAN can access internet through that SOCKS5 proxy server. I tested HTTP Downloader on a client machine and it downloaded files perfectly through that SOCKS5 server. Great job.

One question though. Other download managers like Flareget from https://flareget.com/ allow to enable only one proxy at a time. HTTP Downloader on the other hand allows all the three proxy (HTTP, HTTPS, SOCKS4/SOCKS5) options to be checked simultaneously. So I guess it will try sequentially?

**Comment by erickutcher** [comment]
It won't do it sequentially. And now that I think about it, they all three can't work at the same time either. I'll fix that in the next release.

**Comment by joysan7** [comment]
Please clarify: If I check the option boxes for HTTP proxy as well as SOCKS proxy, which one will be used for downloading.

**Comment by erickutcher** [comment]
They both will download. The HTTP proxy will proxy only HTTP connections, and the SOCKS proxy will end up handling the rest (HTTPS connections).

If you disable the HTTP proxy and just use SOCKS, then the SOCKS proxy will handle both HTTP and HTTPS.
